File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {
    static StringTokenizer st;
    static int[][] seq;
    static int n, m;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        n = nextInt();
        m = nextInt();

        seq = new int[n + 1][n + 1];

        for (int r = 1; r <= n; r++) {
            st = new StringTokenizer(br.readLine());

            for (int c = 1; c <= n; c++) {
                seq[r][c] = nextInt() + seq[r - 1][c] + seq[r][c - 1] - seq[r - 1][c - 1];
            }
        }

        StringBuilder sb = new StringBuilder();

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int rs = nextInt();
            int cs = nextInt();
            int re = nextInt();
            int ce = nextInt();

            int res = seq[re][ce] - seq[re][cs - 1] - seq[rs - 1][ce] + seq[rs - 1][cs - 1];

            sb.append(res).append('\n');
        }

        System.out.println(sb);
        br.close();
    }

    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }
}

File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.kt
import java.io.StreamTokenizer

fun main() = with(StreamTokenizer(System.`in`.bufferedReader())) {
    val readInt = { nextToken(); nval.toInt() }
    val n = readInt()
    val m = readInt()
    val tabel = Array(n + 1) { IntArray(n + 1) }

    repeat(n) { r ->
        repeat(n) { c->
            tabel[r + 1][c + 1] = readInt() + tabel[r + 1][c] + tabel[r][c + 1] - tabel[r][c]
        }
    }

    val bw = System.out.bufferedWriter()

    repeat(m) {
        val rs = readInt()
        val cs = readInt()
        val re = readInt()
        val ce = readInt()

        val res = tabel[re][ce] - tabel[re][cs - 1] - tabel[rs - 1][ce] + tabel[rs - 1][cs - 1]

        bw.append("$res\n")
    }
    bw.flush()
    bw.close()
}